src/sync.c

```diff
@@ -33,44 +33,35 @@ static void dump_ref(uint8_t *ref_buf)
     // log_debug("REF %08X", value);
 }
 
-static void calc_phase(float complex *buf, unsigned int ref, float *out_phase, float *out_slope)
+static void adjust_ref(float complex *buf, float *phases, unsigned int ref)
 {
+    // sync bits (after DBPSK)
+    static const signed char sync[] = {
+        -1, 1, -1, -1, -1, 1, 1
+    };
     float phase, slope;
     float complex sum;
 
     sum = 0;
-    for (int r = 0; r < BLKSZ; r++)
-        sum += buf[ref * BLKSZ + r] * buf[ref * BLKSZ + r];
-    phase = cargf(sum) * 0.5;
-
-    sum = 0;
     for (int r = 1; r < BLKSZ; r++)
     {
         float complex tmp = conjf(buf[ref * BLKSZ + r - 1]) * buf[ref * BLKSZ + r];
         sum += tmp * tmp;
     }
     slope = cargf(sum) * 0.5;
 
-    *out_phase = phase;
-    *out_slope = slope;
-}
-
-static void adjust_ref(float complex *buf, float *phases, unsigned int ref)
-{
-    // sync bits (after DBPSK)
-    static const signed char sync[] = {
-        -1, 1, -1, -1, -1, 1, 1
-    };
-    float phase, slope;
-    calc_phase(buf, ref, &phase, &slope);
-
     if (prev_slope[ref])
         slope = slope * 0.1 + prev_slope[ref] * 0.9;
     prev_slope[ref] = slope;
 
+    sum = 0;
+    for (int r = 0; r < BLKSZ; r++)
+        sum += buf[ref * BLKSZ + r] * buf[ref * BLKSZ + r] * cexpf(-I * 2 * slope * r);
+    phase = cargf(sum) * 0.5;
```

Comment:
The phase calculation now takes place after the slope is calculated, so that the slope can be subtracted out.

I merged `calc_phase` into `adjust_ref` because this method needs to use the averaged slope. (I tried with the raw slope, but that reduced the receiver's weak-signal performance.)